### PR TITLE
Iterative horizontal fusion.

### PR DESCRIPTION
**Pull request by trentlo**
1. Extend horizontal fusion to support non-fusion instructions.
2. Enable iterative optimization for horizontal fusion. After each iteration,
   new horizontal fusion opportunities are exposed because the producers to
   the previously generated horizontally fused instructions will become
   fusion candidates. See `IterativeHorizontalFusion` in the unittest as an example.

### CODE REVIEW

**Comment by trentlo** [comment]
@cheshire could you help to review this PR? Thanks!

**Comment by cheshire** [comment]
BTW I'm currently working on trying to change the calling convention to allow more than N arguments for calls. If successful, that should enable a lot more horizontal fusion opportunities (in many cases I have seen it hits the # of arguments to the kernel boundary)

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc [hunk 1]
```diff
@@ -67,6 +67,25 @@ PrimitiveType GetUniqueOutputTypeOfFusion(const HloInstruction& instr) {
   return first_output_type;
 }
 
+size_t GetInstrCountOfFusible(const HloInstruction& instr) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:70]: Is it possible to reduce the duplication with the same function in the other file?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc [hunk 2]
```diff
@@ -67,6 +67,25 @@ PrimitiveType GetUniqueOutputTypeOfFusion(const HloInstruction& instr) {
...
+  if (instr.opcode() != HloOpcode::kFusion) {
+    return 1;
+  } else {
+    return instr.fused_instruction_count();
+  }
+}
+
+// Creates a kLoop fusion instruction and fuses `fused` into the created
+// fusion instruction.
+HloInstruction* MakeLoopFusionInstruction(HloInstruction* fused) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:80]: Same note regarding reducing duplication. Fusion type could be simply passed as a parameter?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc [hunk 3]
```diff
@@ -493,16 +524,26 @@ StatusOr<bool> HorizontalLoopFusionImpl::Run() {
...
         // Skip; there is just one fused_instr.
         continue;
       }
 
       changed = true;
-      TF_RETURN_IF_ERROR(Fuse(fusions));
+      // Convert fusible into fusion_instrs to simplify the implementation of
+      // `Fuse()`.
+      std::vector<HloInstruction*> fusion_instrs;
+      for (auto instr : fusibles) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:539]: Explicit type preferred

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc [hunk 4]
```diff
@@ -493,16 +524,26 @@ StatusOr<bool> HorizontalLoopFusionImpl::Run() {
     auto consumer = def_to_use_order[i];
     HorizontalLoopFusionImpl::FusionCandidates fusion_candidates(consumer);
     while (true) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:526]: If we are running this in a fixed point, could we remove this while/true loop then?

**Review by cheshire — CHANGES_REQUESTED** [review]

**Review by trentlo — COMMENTED** [review]
Response to the comments. Will revise the code soon.

**Comment by trentlo** [comment]
> BTW I'm currently working on trying to change the calling convention to allow more than N arguments for calls. If successful, that should enable a lot more horizontal fusion opportunities (in many cases I have seen it hits the # of arguments to the kernel boundary)

Cool! As it is a limitation in the CUDA kernel signature, (wondering if) are you going to pack the arguments into a struct and pass the struct?

**Comment by trentlo** [comment]
I've addressed the review comments, please help to take a look again. Thanks!

**Comment by cheshire** [comment]
> wondering if) are you going to pack the arguments into a struct and pass the struct?

This would not work, as the limit is on the argument size and not the number of arguments. I'm experimenting with passing the buffer table directly instead.

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 5]
```diff
@@ -541,5 +541,35 @@ bool IsConsumerTheOnlyNonRootUser(const HloInstruction& instr,
...
+    return 1;
+  } else {
+    return instr.fused_instruction_count();
+  }
+}
+
+absl::InlinedVector<HloInstruction*, 2> GetOutputsOfFusible(
+    const HloInstruction& instr) {
+  if (instr.opcode() != HloOpcode::kFusion) {
+    return {const_cast<HloInstruction*>(&instr)};
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/gpu_fusible.cc:555]: Must we const-cast? Why not return a vector of const pointers?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc (hunk 5 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc:555]: The return type has to align with the return type of `HloInstruction::operands()`, which is of `absl::InlinedVector<HloInstruction*, 2>`. So, returning absl::InlinedVector<const HloInstruction*, 2> will require new allocation of the vector. `const_cast` may be a bit less evil.

Let me know if you have any other ideas?

Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc:555]: `const_cast` seems to be technically UB if modifying methods are then called, right? Allocating 2 pointers on a stack (new vector) should be essentially free, right?

Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Just did some search. It is undefined behavior if the object itself (not the pointer) is indeed const. Then, changing the object is undefined behavior through const_cast.

My previous statement was wrong though--we allocate a new vector (in the stack) anyway. So, let's remove the const_cast. Please help to take a look of the latest [commit](https://github.com/tensorflow/tensorflow/pull/48706/commits/0e0604868fcb74cda63333a6af1adfa3a38b03fb).

**Review by cheshire — APPROVED** [review]

**Pull request merged**